include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -0,0 +1,391 @@
+    return b;
+  }
+
+  /**
+   * @brief Find an available block in the pool of at least `size` bytes, for use on `stream`.
+   * 
+   * Attempts to find a free block that was last used on `stream` to avoid synchronization. If none
+   * is available, it finds a block last used on another stream. In this case, the stream associated
+   * with the found block is synchronized to ensure all asynchronous work on the memory is finished
+   * before it is used on `stream`.
+   * 
+   * @param size The size of the requested allocation, in bytes.
+   * @param stream The stream on which the allocation will be used.
+   * @return block A block with non-null pointer and size >= `size`.
+   */
+  block available_larger_block(size_t size, cudaStream_t stream) {
+    // Try to find a larger block in free list for the same stream
+    auto iter = stream_blocks_.find(stream);
+    if (iter != stream_blocks_.end()) {
+      block b = block_from_stream(iter->second, stream, size, stream);
+      if (b.ptr != nullptr) return b;
+    }
+
+    // nothing in this stream's free list, look for one on another stream
+    auto s = stream_blocks_.begin();
+    while (s != stream_blocks_.end()) {
+      if (s->first != stream) {
+        block b = block_from_stream(s->second, s->first, size, stream);
+        if (b.ptr != nullptr) return b;
+      }
+      ++s;
+    }
+
+    // no larger blocks available on other streams, so grow the pool and create a block
+    size_t grow_size = size_to_grow(size);
+    RMM_EXPECTS(grow_size > 0, rmm::bad_alloc, "Maximum pool size exceeded");
+    return block_from_upstream(grow_size, stream);
+  }
+
+  /**
+   * @brief Splits block `b` if necessary to return a pointer to memory of `size` bytes.
+   *
+   * If the block is split, the remainder is returned to the pool.
+   * 
+   * @param b The block to allocate from.
+   * @param size The size in bytes of the requested allocation.
+   * @param stream The stream on which the allocation will be used.
+   * @return void* The pointer to the allocated memory.
+   */
+  void* allocate_from_block(block const& b, size_t size, cudaStream_t stream)
+  {
+    block const alloc{b.ptr, size, b.is_head};
+
+    if (b.size > size)
+    {
+      block rest{b.ptr + size, b.size - size, false};
+      stream_blocks_[stream].insert(rest);
+    }
+
+    allocated_blocks_.insert(alloc);
+    return reinterpret_cast<void*>(alloc.ptr);
+  }
+
+  /**
+   * @brief Frees the block associated with pointer `p`, returning it to the pool.
+   * 
+   * @param p The pointer to the memory to free.
+   * @param size The size of the memory to free. Must be equal to the original allocation size.
+   * @param stream The stream on which the memory was last used.
+   */
+  void free_block(void *p, size_t size, cudaStream_t stream)
+  {
+    if (p == nullptr) return;
+
+    auto const i = allocated_blocks_.find(block{static_cast<char*>(p)});
+    assert(i != allocated_blocks_.end());
+
+    if (i != allocated_blocks_.end()) {
```

Comment:
I tried removing this conditional, but on my system `set::erase()` *hangs* if you pass it an iterator that is equal to `allocated_blocks_.end()`.  Since we test this function (`DeallocateShouldNotThrow` test), we can't allow it to hang.

Reply:
Ah, that test is problematic since it's testing a situation that results in undefined behavior. Hanging the application is perfectly allowed in the realm of UB.

Reply:
Hmmm, I wrote a bunch of MRs that threw in deallocate. If we had a test like this I wouldn't have done that. So I added this test.

Can you think of another way to enforce it?

Reply:
I also find it strange that set::erase() would hang...

Reply:
But I agree the test is problematic. Some allocators could segfault, for example, and that would be allowed, but difficult for testing.

Reply:
Test removed. Change made.